src/lib/ast.rs

```diff
@@ -0,0 +1,617 @@
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file //
+// if one is included with the Software (each a “Larger Work” to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::io::{Read, Write};
+use std::ops::{Index, IndexMut};
+use std::path::Path;
+
+use dot;
+use lrlex::{build_lex, Lexeme};
+use lrpar::parser;
+use lrtable::{Grammar, Minimiser, TIdx, yacc_to_statetable};
+
+#[derive(Debug)]
+/// Errors raised when parsing a source file.
+pub enum ParseError {
+    /// Io error returned from standard library routine.
+    Io(io::Error),
+    /// File not found.
+    FileNotFound,
+    /// File name had no extension (used to determine which lexer/parser to use).
+    NoFileExtension,
+    /// Lexer could not be built by `lrlex`.
+    BrokenLexer,
+    /// Parser could not be built by `lrpar`.
+    BrokenParser,
+    /// File contained lexical error and could not be lexed.
+    LexicalError,
+    /// File contained syntax error and could not be parsed.
+    SyntaxError,
+}
+
+/// A node identifier for nodes within a particular `Arena`.
+type NodeId = usize;
```

Comment:
Although you probably realise this, it's worth emphasising (since I was caught out by this) that type aliases are weak. In other words, this just says "you can use NodeId and usize interchangeably". It doesn't actually make a new type called NodeId.

[If you want to make a non-interchangeable type, life gets a little harder. lrtable has examples.]